File: c_programs_and_containers/build_blocks/scan_tree/runtime/st_display.c
/* st_display.c - Hierarchical colored layer state display */

#include "st_display.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ANSI color codes */
#define ANSI_GREEN  "\033[32m"
#define ANSI_RED    "\033[31m"
#define ANSI_GREY   "\033[90m"
#define ANSI_BOLD   "\033[1m"
#define ANSI_RESET  "\033[0m"

/* Pre-computed display entry */
typedef struct {
    uint16_t    buf_id;         /* index into buf_descs[] */
    uint8_t     indent;         /* display indentation level */
    const char *title;          /* pointer into path string (relative) */
} st_disp_entry_t;

struct st_display {
    const char      *name;      /* system name */
    st_disp_entry_t *entries;   /* display order, pre-sorted */
    uint16_t         n_entries;
};

/* Count dots in a path */
static int path_depth(const char *path)
{
    int d = 0;
    while (*path) { if (*path == '.') d++; path++; }
    return d;
}

/* Skip N dot-separated segments, return pointer to remainder */
static const char *skip_segments(const char *path, int n)
{
    int d = 0;
    while (*path && d < n) { if (*path == '.') d++; path++; }
    return path;
}

/* Check if child_path is a direct sub-level child of parent_path */
static int is_child_of(const char *parent_path, const char *child_path, int parent_depth)
{
    const char *p = parent_path;
    const char *last_dot = NULL;
    while (*p) { if (*p == '.') last_dot = p; p++; }
    int prefix_len = last_dot ? (int)(last_dot - parent_path) : 0;

    if (prefix_len == 0) return 0;
    if (strncmp(parent_path, child_path, prefix_len) != 0) return 0;
    if (child_path[prefix_len] != '.') return 0;

    int child_depth = path_depth(child_path);
    return (child_depth == parent_depth + 1) ? 1 : 0;
}

/* Recursively add entries in display order */
static void build_recursive(const st_handle_t *h, uint16_t buf_id,
                              const uint16_t *layer_ids, int n_layers,
                              uint8_t indent,
                              st_disp_entry_t *entries, uint16_t *count)
{
    const st_buf_desc_t *bd = &h->desc->buf_descs[buf_id];

    entries[*count].buf_id = buf_id;
    entries[*count].indent = indent;
    entries[*count].title  = skip_segments(bd->path, 1); /* skip root name */
    (*count)++;

    int my_depth = path_depth(bd->path);
    uint8_t my_level = bd->level;

    for (int i = 0; i < n_layers; i++) {
        if (layer_ids[i] == buf_id) continue;
        const st_buf_desc_t *cbd = &h->desc->buf_descs[layer_ids[i]];
        if (cbd->level == my_level &&
            is_child_of(bd->path, cbd->path, my_depth)) {
            build_recursive(h, layer_ids[i], layer_ids, n_layers,
                            indent + 1, entries, count);
        }
    }
}

st_display_t *st_display_init(const st_handle_t *h)
{
    const st_system_desc_t *d = h->desc;

    /* Collect layer buffer ids */
    uint16_t layer_ids[256];
    int n_layers = 0;
    for (uint16_t i = 0; i < d->n_bufs && n_layers < 256; i++) {
        if (d->buf_descs[i].is_layer)
            layer_ids[n_layers++] = i;
    }

    st_display_t *disp = (st_display_t *)calloc(1, sizeof(st_display_t));
    if (!disp) return NULL;

    disp->name = d->name;
    disp->entries = (st_disp_entry_t *)calloc(n_layers, sizeof(st_disp_entry_t));
    if (!disp->entries) { free(disp); return NULL; }
    disp->n_entries = 0;

    if (n_layers == 0) return disp;

    /* Find max level */
    uint8_t max_level = 0;
    for (int i = 0; i < n_layers; i++) {
        if (d->buf_descs[layer_ids[i]].level > max_level)
            max_level = d->buf_descs[layer_ids[i]].level;
    }

    /* Build display order: levels from highest to lowest,
     * root buffers first within each level, children indented */
    for (int lvl = (int)max_level; lvl >= 0; lvl--) {
        /* Find minimum path depth for this level */
        int min_depth = 9999;
        for (int i = 0; i < n_layers; i++) {
            if (d->buf_descs[layer_ids[i]].level != (uint8_t)lvl) continue;
            int dep = path_depth(d->buf_descs[layer_ids[i]].path);
            if (dep < min_depth) min_depth = dep;
        }

        /* Add root buffers of this level */
        for (int i = 0; i < n_layers; i++) {
            if (d->buf_descs[layer_ids[i]].level != (uint8_t)lvl) continue;
            int dep = path_depth(d->buf_descs[layer_ids[i]].path);
            if (dep == min_depth) {
                build_recursive(h, layer_ids[i], layer_ids, n_layers,
                                0, disp->entries, &disp->n_entries);
            }
        }
    }

    return disp;
}

void st_display_destroy(st_display_t *disp)
{
    if (disp) {
        free(disp->entries);
        free(disp);
    }
}

void st_display_tree(const st_display_t *disp, const st_handle_t *h)
{
    printf(ANSI_BOLD "=== %s ===" ANSI_RESET "\n", disp->name);

    for (uint16_t i = 0; i < disp->n_entries; i++) {
        const st_disp_entry_t *e = &disp->entries[i];
        const st_buf_desc_t *bd = &h->desc->buf_descs[e->buf_id];
        const st_layer_rt_t *l = &h->layer_bufs[bd->buf_index];

        for (uint8_t j = 0; j < e->indent; j++) printf("  ");

        printf(ANSI_BOLD "%s" ANSI_RESET " [", e->title);

        for (uint16_t j = 0; j < bd->size; j++) {
            if (j > 0) printf(" ");
            int8_t s = l->states[j];
            if (s == 1)
                printf(ANSI_GREEN "T" ANSI_RESET);
            else if (s == 0)
                printf(ANSI_RED "F" ANSI_RESET);
            else
                printf(ANSI_GREY "N" ANSI_RESET);
        }
        printf("]\n");
    }
}

File: luajit_programs_and_containers/building_blocks/scan_tree_luajit/st_display.lua
--[[
    st_display.lua - Hierarchical Colored Layer State Display (LuaJIT)

    Port of st_display.c: builds a cached display hierarchy from the
    handle's buffer descriptors, then prints colored state trees.
--]]

local M = {}

-- ANSI color codes
local GREEN = "\027[32m"
local RED   = "\027[31m"
local GREY  = "\027[90m"
local BOLD  = "\027[1m"
local RESET = "\027[0m"

-- Count dots in a path
local function path_depth(path)
    local d = 0
    for _ in path:gmatch("%.") do d = d + 1 end
    return d
end

-- Skip N dot-separated segments, return remainder
local function skip_segments(path, n)
    local d = 0
    local i = 1
    while i <= #path and d < n do
        if path:sub(i, i) == "." then d = d + 1 end
        i = i + 1
    end
    return path:sub(i)
end

-- Check if child_path is a direct sub-level child of parent_path
local function is_child_of(parent_path, child_path, parent_depth)
    -- Find the prefix up to the last dot in parent_path
    local last_dot = nil
    for i = 1, #parent_path do
        if parent_path:sub(i, i) == "." then last_dot = i end
    end
    local prefix_len = last_dot and (last_dot - 1) or 0
    if prefix_len == 0 then return false end

    local prefix = parent_path:sub(1, prefix_len)
    if child_path:sub(1, prefix_len) ~= prefix then return false end
    if child_path:sub(prefix_len + 1, prefix_len + 1) ~= "." then return false end

    return path_depth(child_path) == parent_depth + 1
end

-- =====================================================================
-- Display object
-- =====================================================================

local Display = {}
Display.__index = Display

-- Build display hierarchy recursively
local function build_recursive(handle, buf_id, layer_ids, indent, entries)
    local bd = handle.desc.buf_descs[buf_id + 1]
    entries[#entries + 1] = {
        buf_id = buf_id,
        indent = indent,
        title  = skip_segments(bd.path, 1),
    }

    local my_depth = path_depth(bd.path)
    local my_level = bd.level

    for _, lid in ipairs(layer_ids) do
        if lid ~= buf_id then
            local cbd = handle.desc.buf_descs[lid + 1]
            if cbd.level == my_level and is_child_of(bd.path, cbd.path, my_depth) then
                build_recursive(handle, lid, layer_ids, indent + 1, entries)
            end
        end
    end
end

function Display.new(handle)
    local self = setmetatable({}, Display)
    local d = handle.desc

    self.name = d.name

    -- Collect layer buffer ids (0-based)
    local layer_ids = {}
    for i = 1, #d.buf_descs do
        if d.buf_descs[i].is_layer then
            layer_ids[#layer_ids + 1] = i - 1
        end
    end

    self.entries = {}
    if #layer_ids == 0 then return self end

    -- Find max level
    local max_level = 0
    for _, lid in ipairs(layer_ids) do
        local lvl = d.buf_descs[lid + 1].level
        if lvl > max_level then max_level = lvl end
    end

    -- Build display order: levels from highest to lowest
    for lvl = max_level, 0, -1 do
        -- Find minimum path depth for this level
        local min_depth = 9999
        for _, lid in ipairs(layer_ids) do
            if d.buf_descs[lid + 1].level == lvl then
                local dep = path_depth(d.buf_descs[lid + 1].path)
                if dep < min_depth then min_depth = dep end
            end
        end

        -- Add root buffers of this level
        for _, lid in ipairs(layer_ids) do
            if d.buf_descs[lid + 1].level == lvl then
                if path_depth(d.buf_descs[lid + 1].path) == min_depth then
                    build_recursive(handle, lid, layer_ids, 0, self.entries)
                end
            end
        end
    end

    return self
end

function Display:print_tree(handle)
    io.write(BOLD .. "=== " .. self.name .. " ===" .. RESET .. "\n")

    for _, e in ipairs(self.entries) do
        local bd = handle.desc.buf_descs[e.buf_id + 1]
        local l = handle.layer_bufs[bd.buf_index]

        -- Indentation
        for _ = 1, e.indent do io.write("  ") end

        io.write(BOLD .. e.title .. RESET .. " [")

        for j = 0, bd.size - 1 do
            if j > 0 then io.write(" ") end
            local s
            if l.not_active[j] ~= 0 then
                s = -1
            elseif l.value[j] ~= 0 then
                s = 1
            else
                s = 0
            end

            if s == 1 then
                io.write(GREEN .. "T" .. RESET)
            elseif s == 0 then
                io.write(RED .. "F" .. RESET)
            else
                io.write(GREY .. "N" .. RESET)
            end
        end
        io.write("]\n")
    end
end

M.Display = Display

return M
